traffic_ops/traffic_ops_golang/auth/authenticate.go

```diff
@@ -0,0 +1,201 @@
+	DKLen:     64}
+
+// DerivePassword uses the golang.org/x/crypto package to
+// return an encrypted password that is compatible with the
+// Perl CPAN library Crypt::ScryptKDF for backward compatibility
+// to authenticate through the Perl API the same way.
+// See: http://cpansearch.perl.org/src/MIK/Crypt-ScryptKDF-0.010/lib/Crypt/ScryptKDF.pm
+func DerivePassword(password string) (string, error) {
+	var salt []byte
+	var err error
+	salt, err = generateSalt(64)
+	if err != nil {
+		return "", err
+	}
+	key, err := scrypt.Key([]byte(password), salt, DefaultParams.N, DefaultParams.R, DefaultParams.P, DefaultParams.DKLen)
+	if err != nil {
+		return "", err
+	}
+	nStr := strconv.Itoa(DefaultParams.N)
+	if err != nil {
+		return "", err
+	}
+	rStr := strconv.Itoa(DefaultParams.R)
+	pStr := strconv.Itoa(DefaultParams.P)
+	saltBase64 := base64.StdEncoding.EncodeToString(salt)
+	keyBase64 := base64.StdEncoding.EncodeToString(key)
+
+	// The SCRYPT prefix is added because the Mojolicious Perl library adds this as a
+	// prefix to every password in the database.  So it's added for compatibility.
+	scryptPass := []string{DefaultParams.Algorithm, nStr, rStr, pStr, saltBase64, keyBase64}
+
+	return strings.Join(scryptPass, KEY_DELIM), nil
+}
+
+// VerifyPassword parses the original Derived Key (DK) from the SCRYPT password
+// so that it can compare that with the password/scriptPassword param
+func VerifyPassword(password string, scryptPassword string) error {
+
+	scomp, err := parseScrypt(scryptPassword)
+	if err != nil {
+		return err
+	}
+
+	keylenBytes := len(scryptPassword) - DefaultParams.DKLen
+	if keylenBytes < 1 {
+		return errors.New("Invalid targetKey length")
+	}
+	// scrypt the cleartext password with the same parameters and salt
+	tmpDK, err := scrypt.Key([]byte(password),
+		[]byte(scomp.Salt),
+		scomp.N, // Must be a power of 2 greater than 1
+		scomp.R,
+		scomp.P, // r*p must be < 2^30
+		DefaultParams.DKLen)
+	if err != nil {
+		return err
+	}
+	// Compare the Derived Key from the SCRYPT password
+	if subtle.ConstantTimeCompare(scomp.DK, tmpDK) == 1 {
+		return nil
+	}
+
+	return err
+}
+
+func parseScrypt(scryptPassword string) (SCRYPTComponents, error) {
+	sh := strings.Split(scryptPassword, ":")
+
+	var err error
+	var scomp SCRYPTComponents
+	if scryptPassword == "" {
+		return scomp, errors.New("scrypt password is required")
+	}
+
+	// Algorithm
+	scomp.Algorithm = sh[0]
+	if scomp.Algorithm == "" {
+		return scomp, errors.New("Algorithm was not defined")
+	}
+
+	// N
+	n := sh[1]
+	if n == "" {
+		return scomp, errors.New("N was not defined")
+	}
+	var nInt int
+	nInt, err = strconv.Atoi(n)
+	if err != nil {
+		return scomp, errors.New(fmt.Sprintf("%v i=%d, type: %T\n", err, nInt, nInt))
+	}
+	scomp.N = nInt
+
+	// R
+	r := sh[2]
+	if r == "" {
+		return scomp, errors.New("r was not defined")
+	}
+
+	scomp.R, err = strconv.Atoi(r)
+	if err != nil {
+		return scomp, errors.New(fmt.Sprintf("i=%d, type: %T\n", scomp.R, scomp.R))
```

Comment:
Minor suggestion: could be `fmt.Errorf("i=%d, type: %T", scomp.R, scomp.R)`. Suggest removing newline, errors don't usually have newlines

Reply:
good to know, wasn't aware of fmt.Errorf